packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
+    if (newValue == _highlightColor)
+      return;
+    _highlightColor = newValue;
+    notifyListeners();
+  }
+
+  TextRange? get highlightedRange => _highlightedRange;
+  TextRange? _highlightedRange;
+  set highlightedRange(TextRange? newValue) {
+    if (newValue == _highlightedRange)
+      return;
+    _highlightedRange = newValue;
+    notifyListeners();
+  }
+
+  /// Controls how tall the selection highlight boxes are computed to be.
+  ///
+  /// See [ui.BoxHeightStyle] for details on available styles.
+  ui.BoxHeightStyle get selectionHeightStyle => _selectionHeightStyle;
+  ui.BoxHeightStyle _selectionHeightStyle = ui.BoxHeightStyle.tight;
+  set selectionHeightStyle(ui.BoxHeightStyle value) {
+    assert(value != null);
+    if (_selectionHeightStyle == value)
+      return;
+    _selectionHeightStyle = value;
+    notifyListeners();
+  }
+
+  /// Controls how wide the selection highlight boxes are computed to be.
+  ///
+  /// See [ui.BoxWidthStyle] for details on available styles.
+  ui.BoxWidthStyle get selectionWidthStyle => _selectionWidthStyle;
+  ui.BoxWidthStyle _selectionWidthStyle = ui.BoxWidthStyle.tight;
+  set selectionWidthStyle(ui.BoxWidthStyle value) {
+    assert(value != null);
+    if (_selectionWidthStyle == value)
+      return;
+    _selectionWidthStyle = value;
+    notifyListeners();
+  }
+
+  @override
+  void paint(Canvas canvas, Size size, RenderEditable renderEditable) {
+    final TextRange? range = highlightedRange;
+    final Color? color = highlightColor;
+    if (range == null || color == null || range.isCollapsed) {
+      return;
+    }
+
+    highlightPaint.color = color;
+    final List<TextBox> boxes = renderEditable._textPainter.getBoxesForSelection(
+      TextSelection(baseOffset: range.start, extentOffset: range.end),
+      boxHeightStyle: selectionHeightStyle,
+      boxWidthStyle: selectionWidthStyle
+    );
+
+    for (final TextBox box in boxes)
+      canvas.drawRect(box.toRect().shift(renderEditable._paintOffset), highlightPaint);
+  }
+
+  @override
+  bool shouldRepaint(RenderEditablePainter? oldDelegate) {
+    if (identical(oldDelegate, this))
+      return false;
+    if (oldDelegate == null)
+      return highlightColor != null && highlightedRange != null;
+    return oldDelegate is! _TextHighlightPainter
+        || oldDelegate.highlightColor != highlightColor
+        || oldDelegate.highlightedRange != highlightedRange
+        || oldDelegate.selectionHeightStyle != selectionHeightStyle
+        || oldDelegate.selectionWidthStyle != selectionWidthStyle;
+  }
+}
+
+class _FloatingCursorPainter extends RenderEditablePainter {
+  _FloatingCursorPainter(this.caretPaintCallback);
+
+  bool get shouldPaint => _shouldPaint;
+  bool _shouldPaint = true;
+  set shouldPaint(bool value) {
+    if (shouldPaint == value)
+      return;
+    _shouldPaint = value;
+    notifyListeners();
+  }
+
+  CaretChangedHandler caretPaintCallback;
+
+  bool showRegularCaret = false;
```

Comment:
Does not trigger a repaint because this is floating cursor stuff. Might need a bit of audit when we move the caret painting stuff to a different place.